### DISCUSSION TITLE
Error compile on ESO32 Dev Module

### DISCUSSION THREAD

**Discussion by babos** [Hieromon/AutoConnect#422] [How can I?] [answered]
Hi there,
i tried some examples but i can't compile the autoconnect library. Trying to include only the <Autoconnect.h> library I have compilation errors:

``
#include <WebServer.h>
#include <WiFi.h>
#include <AutoConnect.h>

void setup() {
  // put your setup code here, to run once:

}

void loop() {
  // put your main code here, to run repeatedly:

}
``
Result:
``
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp: In member function 'virtual PageElement* AutoConnect::_setupPage(String&)':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1392:35: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_STAT));
                                   ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1420:40: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_CONFIGNEW));
                                        ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1441:41: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_CONNECTING));
                                         ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1455:40: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_OPENCREDT));
                                        ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1470:38: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_DISCONN));
                                      ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1481:40: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_RESETTING));
                                        ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1490:37: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR("{{RESULT}}"));
                                     ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1496:38: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_SUCCESS));
                                      ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:1517:35: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
     elm->setMold(FPSTR(_PAGE_FAIL));
                                   ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectPage.cpp:20:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp: In member function 'void AutoConnect::_startWebServer()':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:1137:20: error: 'class PageBuilder' has no member named 'transferEncoding'
     _responsePage->transferEncoding(PageBuilder::TransferEncoding_t::AUTOCONNECT_HTTP_TRANSFER);
                    ^~~~~~~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:1137:50: error: 'PageBuilder::TransferEncoding_t' has not been declared
     _responsePage->transferEncoding(PageBuilder::TransferEncoding_t::AUTOCONNECT_HTTP_TRANSFER);
                                                  ^~~~~~~~~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp: In member function 'void AutoConnect::_handleNotFound()':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:1249:129: error: no matching function for call to 'PageElement::PageElement(const __FlashStringHelper*, <brace-enclosed initializer list>)'
       PageElement page404(FPSTR(_PAGE_404), { { F("HEAD"), std::bind(&AutoConnect::_token_HEAD, this, std::placeholders::_1) } });
                                                                                                                                 ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:10:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:123:3: note: candidate: 'PageElement::PageElement(const char*, TokenVT)'
   PageElement(const char* mold, TokenVT source) : _mold(mold), _source(source) {}
   ^~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:123:3: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:122:12: note: candidate: 'PageElement::PageElement(const char*)'
   explicit PageElement(const char* mold) : _mold(mold), _source(std::vector<TokenSourceST>()) {}
            ^~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:122:12: note:   candidate expects 1 argument, 2 provided
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:121:3: note: candidate: 'PageElement::PageElement()'
   PageElement() : _mold(nullptr) {}
   ^~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:121:3: note:   candidate expects 0 arguments, 2 provided
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:119:7: note: candidate: 'PageElement::PageElement(const PageElement&)'
 class PageElement {
       ^~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:119:7: note:   candidate expects 1 argument, 2 provided
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:1251:25: error: no matching function for call to 'PageElement::build(String&)'
       page404.build(html);
                         ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:10:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:128:17: note: candidate: 'String PageElement::build()'
   String        build(void);
                 ^~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:128:17: note:   candidate expects 0 arguments, 1 provided
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:129:17: note: candidate: 'static String PageElement::build(const char*, TokenVT, PageArgument&)'
   static String build(const char* mold, TokenVT tokenSource, PageArgument& args);
                 ^~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:129:17: note:   candidate expects 3 arguments, 1 provided
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp: In member function 'void AutoConnect::_purgePages()':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.cpp:1466:18: error: 'class PageBuilder' has no member named 'clearElements'; did you mean 'clearElement'?
   _responsePage->clearElements();
                  ^~~~~~~~~~~~~
                  clearElement
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp: In constructor 'AutoConnectAux::AutoConnectAux(const String&, const String&, bool, AutoConnectElementVT)':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp:90:33: error: 'PageBuilder::TransferEncoding_t' has not been declared
   transferEncoding(PageBuilder::TransferEncoding_t::AUTOCONNECT_HTTP_TRANSFER);
                                 ^~~~~~~~~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp:90:3: error: 'transferEncoding' was not declared in this scope
   transferEncoding(PageBuilder::TransferEncoding_t::AUTOCONNECT_HTTP_TRANSFER);
   ^~~~~~~~~~~~~~~~
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp:90:3: note: suggested alternative: 'transferEncoding_t'
   transferEncoding(PageBuilder::TransferEncoding_t::AUTOCONNECT_HTTP_TRANSFER);
   ^~~~~~~~~~~~~~~~
   transferEncoding_t
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp: In member function 'PageElement* AutoConnectAux::_setupPage(const String&)':
/Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp:558:36: error: no matching function for call to 'PageElement::setMold(const __FlashStringHelper*)'
       elm->setMold(FPSTR(_PAGE_AUX));
                                    ^
In file included from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnect.h:30,
                 from /Users/babos/Documents/Arduino/libraries/AutoConnect/src/AutoConnectAux.cpp:10:
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note: candidate: 'void PageElement::setMold(const char*)'
   void          setMold(const char* mold) { _mold = mold; }
                 ^~~~~~~
/Users/babos/Documents/Arduino/libraries/PageBuilder/src/PageBuilder.h:130:17: note:   no known conversion for argument 1 from 'const __FlashStringHelper*' to 'const char*'
Più di una libreria trovata per "SD.h"
Usata: /Users/babos/Library/Arduino15/packages/esp32/hardware/esp32/2.0.1-RC1/libraries/SD
Non usata: /Applications/Arduino.app/Contents/Java/libraries/SD
Più di una libreria trovata per "WiFi.h"
Usata: /Users/babos/Library/Arduino15/packages/esp32/hardware/esp32/2.0.1-RC1/libraries/WiFi
Non usata: /Applications/Arduino.app/Contents/Java/libraries/WiFi
exit status 1
Errore durante la compilazione per la scheda ESP32 Dev Module.

``
do you have an idea what I can do?
Best regards

Silvio

**Comment by Hieromon** [comment, marked as answer]
You are using a version of PageBuilder that is not compatible with AutoConnect v1.2.3 or later.
Refs: [README.md](https://github.com/Hieromon/AutoConnect#123-jan-3-2021), [Change log](https://hieromon.github.io/AutoConnect/changelog.html#123-jan-3-2021) of the Documentation.

**Reply by babos** [reply]
thank you boss and best regards

**Reply by babos** [reply]
Hi Hieromon,
I wanted to thank you. I had the previous version of PageBuilder; updating it I was able to compile and everything works fine apart from the RESET which does not bring me back to the initial state.
Still great work! Well done.
Best Regards
Silvio
